### ISSUE TITLE
[FORMATTING] GO Statement Before Procedure Declaration

### ISSUE THREAD

**Issue by grantwforsythe** [sql-formatter-org/sql-formatter#606]
**Input data**

```tsql
SET
NOCOUNT ON;

GO

CREATE OR ALTER PROCEDURE dbo.sp_say_hi (@Name VARCHAR(25) = 'World')
AS BEGIN
    PRINT CONCAT('Hello, ', @Name);
END;

```

**Expected Output**

```tsql
SET
NOCOUNT ON;

GO

CREATE
OR ALTER
PROCEDURE dbo.sp_say_hi (@Name VARCHAR(25) = 'World')
AS BEGIN
    PRINT CONCAT('Hello, ', @Name);
END;
```

The `GO` statement should be on it's own line.

**Actual Output**

```tsql
SET
NOCOUNT ON;

GO CREATE
OR ALTER
PROCEDURE dbo.sp_say_hi (@Name VARCHAR(25) = 'World') AS BEGIN
    PRINT CONCAT('Hello, ', @Name);
END;
```
When executed on the database a fatal error occurs (`Incorrect syntax was encountered while CREATE was being parsed`) because of the `GO`.

**Usage**

- How are you calling / using the library?
The library is being called from a Git hook with the below config:
  ```json
  {
    "language": "tsql",
    "tab width": 4,
    "useTabs": false,
    "keywordCase": "upper",
    "indentStyle": "standard",
    "logicalOperatorNewline": "before",
    "expressionWidth": 80,
    "linesBetweenQueries": 2,
    "denseOperators": false,
    "newlineBeforeSemicolon": false
  }
  ```
- What SQL language(s) does this apply to?
Transact-SQL/T-SQL
- Which SQL Formatter version are you using?
Latest (`v12.2.2`)

**Comment by nene** [comment]
Thanks, the fix is now released in 12.2.2

**Comment by nene** [comment]
Thanks for reporting.

The problem here is that this is not really about how the `GO` statement is supported, but rather how all other statements that can come after `GO` are supported. I threw in some support for `CREATE/ALTER/DROP DATABASE` and `USE` statements. But that's only a needle in the haystack - a lot more is missing. I personally have zero experience with the Microsoft SQL stack. Somebody contributed the Transact-SQL support a long time ago, but turns out that he simply copy-basted the chapter titles from the documentation as syntax, which doesn't really work :P So as a result the Transact-SQL support in SQL Formatter is really pretty crappy. Additionally I've come to understand that there really are multiple SQL languages gathered under the Transact-SQL label. The primary one is for SQL Server, but then there's also Azure and various other things. And I'm not too keen on working on improving this as I find the syntax diagrams in MS docs to be kinda hard to comprehend.

I should point out though that you haven't selected Transact-SQL as the SQL dialect you're formatting (see also the [FAQ](https://github.com/sql-formatter-org/sql-formatter?tab=readme-ov-file#parse-error-unexpected--at-line-)):

<img width="187" alt="Image" src="https://github.com/user-attachments/assets/9208d4eb-a0a9-4603-9234-68635aeeca02" />

Given the state of Transact-SQL support, this might actually make some things worse.

**Comment by Jackesque** [comment]
After changing the dialect, the result did change, just not in any significantly helpful way. Thanks for the reply though.

The result using Transact-SQL:

```sql
USE master;

GO DROP
DATABASE test;

GO CREATE
DATABASE test;

GO USE test;

GO
```

Desired output:

```sql
USE master;

GO
DROP DATABASE test;

GO
CREATE DATABASE test;

GO
USE test;

GO
```

**Comment by nene** [comment]
The fix for the latter has now been released in [15.5.2](https://github.com/sql-formatter-org/sql-formatter/releases/tag/v15.5.2)
